// pixi.js/dist/packages/html-source.js
/*!
 * PixiJS - v8.19.0
 * Compiled Thu, 04 Jun 2026 08:22:17 UTC
 *
 * PixiJS is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 */
this.PIXI = this.PIXI || {};
var html_source_js = (function (exports) {
    'use strict';

    "use strict";


    function isElementImage(resource) {
      const ElementImageCtor = globalThis.ElementImage;
      return !!ElementImageCtor && resource instanceof ElementImageCtor;
    }
    class ElementImageSource extends PIXI.TextureSource {
      /**
       * @param options - Options for creating the snapshot source. `resource` is required.
       * @example
       * ```ts
       * const source = new ElementImageSource({
       *     resource: snapshot, // an ElementImage from captureElementImage()
       *     autoClose: true,    // close the snapshot when this source is destroyed
       * });
       * ```
       */
      constructor(options) {
        if (!options.resource) {
          throw new Error("[ElementImageSource] resource is required.");
        }
        super(options);
        /** The upload method for this texture. */
        this.uploadMethodId = "html";
        /** Snapshots are immutable, so the source is ready as soon as it is constructed. */
        this.isReady = true;
        this._autoClose = options.autoClose === true;
      }
      /**
       * Tests whether a resource is an {@link ElementImage} snapshot, used during automatic
       * source detection (`Texture.from`, `TextureSource.from`).
       * @param resource - The resource to test.
       * @returns `true` if this source can handle the resource.
       */
      static test(resource) {
        return isElementImage(resource);
      }
      /**
       * The width of the snapshot in pixels, rounded up.
       * @example
       * ```ts
       * const source = new ElementImageSource({ resource: snapshot });
       *
       * console.log(source.resourceWidth, source.resourceHeight);
       * ```
       */
      get resourceWidth() {
        return Math.ceil(this.resource.width);
      }
      /**
       * The height of the snapshot in pixels, rounded up.
       * @example
       * ```ts
       * const source = new ElementImageSource({ resource: snapshot });
       *
       * console.log(source.resourceWidth, source.resourceHeight);
       * ```
       */
      get resourceHeight() {
        return Math.ceil(this.resource.height);
      }
      /**
       * Destroys the underlying texture source. When {@link ElementImageSourceOptions.autoClose}
       * was set, also calls {@link ElementImage.close} on the snapshot.
       * @example
       * ```ts
       * const source = new ElementImageSource({ resource: snapshot });
       *
       * source.destroy();
       * snapshot.close(); // release the snapshot yourself unless autoClose was set
       * ```
       */
      destroy() {
        const snapshot = this.resource;
        super.destroy();
        if (this._autoClose && snapshot) {
          snapshot.close();
        }
      }
    }
    /**
     * Registers the source with the {@link extensions} system at the lowest texture-source
     * priority, so automatic detection only falls back to it when no other built-in source
     * claims the resource.
     */
    ElementImageSource.extension = {
      type: PIXI.ExtensionType.TextureSource,
      priority: -10
    };

    "use strict";

    function ensureAllocated(gl, glTexture, target, width, height) {
      if (glTexture.width === width && glTexture.height === height) {
        return;
      }
      gl.texImage2D(
        target,
        0,
        glTexture.internalFormat,
        width,
        height,
        0,
        glTexture.format,
        glTexture.type,
        null
      );
      glTexture.width = width;
      glTexture.height = height;
    }
    const glUploadHTMLResource = {
      extension: {
        type: PIXI.ExtensionType.TextureUploaderWebGL,
        name: "html"
      },
      id: "html",
      upload(source, glTexture, gl, _webGLVersion, targetOverride) {
        var _a;
        const upload = gl.texElementImage2D;
        if (!upload) {
          throw new Error(
            // eslint-disable-next-line max-len
            "[HTMLSource] WebGLRenderingContext.texElementImage2D is not available. Enable the browser HTML-in-Canvas API before using HTMLSource."
          );
        }
        const target = targetOverride != null ? targetOverride : glTexture.target;
        const textureWidth = source.pixelWidth;
        const textureHeight = source.pixelHeight;
        if (!source.isReady) {
          ensureAllocated(gl, glTexture, target, textureWidth, textureHeight);
          (_a = source.requestPaint) == null ? void 0 : _a.call(source);
          return;
        }
        upload.call(
          gl,
          target,
          0,
          glTexture.internalFormat,
          glTexture.format,
          glTexture.type,
          source.resource
        );
        glTexture.width = textureWidth;
        glTexture.height = textureHeight;
      }
    };

    "use strict";

    const gpuUploadHTMLResource = {
      extension: {
        type: PIXI.ExtensionType.TextureUploaderWebGPU,
        name: "html"
      },
      type: "html",
      upload(source, gpuTexture, gpu, originZOverride = 0) {
        var _a;
        const queue = gpu.device.queue;
        const copyElementImageToTexture = queue.copyElementImageToTexture;
        if (!copyElementImageToTexture) {
          throw new Error(
            // eslint-disable-next-line max-len
            "[HTMLSource] GPUQueue.copyElementImageToTexture is not available. Enable the browser HTML-in-Canvas API before using HTMLSource."
          );
        }
        if (!source.isReady) {
          (_a = source.requestPaint) == null ? void 0 : _a.call(source);
          return;
        }
        const premultipliedAlpha = source.alphaMode === "premultiply-alpha-on-upload";
        const destination = {
          texture: gpuTexture,
          origin: { x: 0, y: 0, z: originZOverride },
          premultipliedAlpha
        };
        const width = Math.min(gpuTexture.width, source.pixelWidth);
        const height = Math.min(gpuTexture.height, source.pixelHeight);
        copyElementImageToTexture.call(queue, source.resource, width, height, destination);
      }
    };

    "use strict";
    var __defProp = Object.defineProperty;
    var __getOwnPropSymbols = Object.getOwnPropertySymbols;
    var __hasOwnProp = Object.prototype.hasOwnProperty;
    var __propIsEnum = Object.prototype.propertyIsEnumerable;
    var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
    var __spreadValues = (a, b) => {
      for (var prop in b || (b = {}))
        if (__hasOwnProp.call(b, prop))
          __defNormalProp(a, prop, b[prop]);
      if (__getOwnPropSymbols)
        for (var prop of __getOwnPropSymbols(b)) {
          if (__propIsEnum.call(b, prop))
            __defNormalProp(a, prop, b[prop]);
        }
      return a;
    };


    function isCanvas(resource) {
      return !!globalThis.HTMLCanvasElement && resource instanceof HTMLCanvasElement;
    }
    const _HTMLSource = class _HTMLSource extends PIXI.TextureSource {
      /**
       * @param options - Options for creating the HTML source. `resource` is required.
       * @example
       * ```ts
       * const source = new HTMLSource({
       *     resource: domElement,   // a direct child of the Pixi canvas
       *     autoUpdate: true,       // track browser repaints (default)
       *     autoRequestPaint: true, // request one initial paint (default)
       * });
       * ```
       */
      constructor(options) {
        var _a;
        options = __spreadValues(__spreadValues({}, _HTMLSource.defaultOptions), options);
        if (!options.resource) {
          throw new Error("[HTMLSource] resource is required.");
        }
        super(options);
        /** The upload method for this texture. */
        this.uploadMethodId = "html";
        const canvas = (_a = options.canvas) != null ? _a : this._inferCanvas(options.resource);
        if (!canvas) {
          throw new Error(
            // eslint-disable-next-line max-len
            "[HTMLSource] Could not determine the owning canvas. Append the element to the canvas before constructing this source, or pass the `canvas` option."
          );
        }
        if (options.resource.parentElement !== canvas) {
          throw new Error(
            // eslint-disable-next-line max-len
            "[HTMLSource] resource must be a direct child of the owning canvas. Append the element to the canvas before constructing this source."
          );
        }
        this.canvas = canvas;
        this._autoUpdate = options.autoUpdate !== false;
        this._onPaintBound = this._onPaint.bind(this);
        this._isReady = !this._autoUpdate || !canvas.requestPaint;
        if (options.autoLayout !== false) {
          canvas.setAttribute("layoutsubtree", "");
        }
        if (this._autoUpdate) {
          canvas.addEventListener("paint", this._onPaintBound);
        }
        if (options.autoRequestPaint !== false) {
          this.requestPaint();
        }
      }
      /**
       * Tests whether a resource should be handled by `HTMLSource` during automatic source
       * detection (`Texture.from`, `TextureSource.from`). Deliberately strict: only generic HTML
       * elements pass. Image, video, and canvas elements are rejected because they have
       * dedicated, faster sources; snapshots are handled by {@link ElementImageSource}.
       * @param resource - The resource to test.
       * @returns `true` if this source can handle the resource.
       */
      static test(resource) {
        return !!globalThis.HTMLElement && resource instanceof HTMLElement && !(resource instanceof HTMLImageElement) && !(resource instanceof HTMLVideoElement) && !(resource instanceof HTMLCanvasElement);
      }
      /**
       * `true` once the owning canvas has produced an initial paint snapshot, so the texture has
       * real pixels. Non-auto-updating sources are ready immediately.
       * @example
       * ```ts
       * const source = new HTMLSource({ resource: domElement });
       *
       * if (!source.isReady)
       * {
       *     // The first paint has not landed yet — the texture is still blank.
       * }
       * ```
       */
      get isReady() {
        return this._isReady;
      }
      /**
       * Request a `paint` event from the owning canvas. Call this every frame to keep an animated
       * or frequently-changing element in sync with the rendered texture.
       * @returns `true` if the request was made, `false` when the browser lacks the experimental
       * `requestPaint` API or there is no owning canvas.
       * @example
       * ```ts
       * const source = new HTMLSource({ resource: clock, autoRequestPaint: false });
       *
       * app.ticker.add(() => {
       *     clock.textContent = new Date().toLocaleTimeString();
       *     source.requestPaint();
       * });
       * ```
       */
      requestPaint() {
        var _a;
        if (!((_a = this.canvas) == null ? void 0 : _a.requestPaint)) {
          return false;
        }
        this.canvas.requestPaint();
        return true;
      }
      /**
       * Detaches the `paint` listener from the owning canvas and destroys the underlying texture
       * source.
       * @example
       * ```ts
       * const source = new HTMLSource({ resource: domElement });
       *
       * source.destroy();
       * ```
       */
      destroy() {
        if (this.canvas && this._autoUpdate) {
          this.canvas.removeEventListener("paint", this._onPaintBound);
        }
        this.canvas = null;
        super.destroy();
      }
      /** Width in real pixels (`offsetWidth`). Use {@link width} for CSS pixels. */
      get resourceWidth() {
        return this.resource.offsetWidth || 1;
      }
      /** Height in real pixels (`offsetHeight`). Use {@link height} for CSS pixels. */
      get resourceHeight() {
        return this.resource.offsetHeight || 1;
      }
      _inferCanvas(resource) {
        return isCanvas(resource.parentElement) ? resource.parentElement : null;
      }
      _onPaint(event) {
        const changedElements = event.changedElements;
        if ((changedElements == null ? void 0 : changedElements.length) && !changedElements.includes(this.resource)) {
          return;
        }
        this._isReady = true;
        this.update();
      }
    };
    /**
     * Registers the source with the {@link extensions} system at the lowest texture-source
     * priority, so automatic detection only falls back to it when no other built-in source
     * claims the resource.
     */
    _HTMLSource.extension = {
      type: PIXI.ExtensionType.TextureSource,
      priority: -10
    };
    /**
     * The default options applied to every {@link HTMLSource}, merged with the options passed
     * to the constructor.
     * @example
     * ```ts
     * // Make every HTMLSource opt out of automatic repaint tracking by default.
     * HTMLSource.defaultOptions.autoUpdate = false;
     * ```
     */
    _HTMLSource.defaultOptions = {
      autoLayout: true,
      autoUpdate: true,
      autoRequestPaint: true
    };
    let HTMLSource = _HTMLSource;

    "use strict";

    "use strict";

    "use strict";
    PIXI.extensions.add(
      HTMLSource,
      ElementImageSource,
      glUploadHTMLResource,
      gpuUploadHTMLResource
    );

    "use strict";

    exports.ElementImageSource = ElementImageSource;
    exports.HTMLSource = HTMLSource;
    exports.glUploadHTMLResource = glUploadHTMLResource;
    exports.gpuUploadHTMLResource = gpuUploadHTMLResource;

    return exports;

})({});
Object.assign(this.PIXI, html_source_js);
